File: matlab/matrixDataProcess.m
% data proccessing for white noise

close all
clear all
format long

%%%%%%%%%%%%%%%%%%%%%%%


Fs=38e6; %Sample freq
wd=10; % window duration [us] for the correlation purpose
path="/dev/shm/";



fid=fopen(path+"fdataR.dat",'rb');
dataR=fread(fid,'float');


fid=fopen(path+"fdataI.dat",'rb');
dataI=fread(fid,'float');

dataR = dataR(1:2*Fs);
dataI = dataI(1:2*Fs);
dataC=complex(dataR,dataI); % received complex data

clear dataR
clear dataI

lenDRaw=length(dataC(:,1));

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%Calibration processing



calibrationOffset = 0.6 *Fs ; % conversion from time to samples



% signal calibration
offset =offsetcalc(dataC (1:0.6*Fs),Fs); % received samples offset due to hardware & software lag [samples]floor(0.447352158*Fs)

%snr = SNRcalc(dataC (1:0.6*Fs),Fs,offset);
dataC  = dataC (offset+1:offset +1*Fs);



lenD=length(dataC(:,1));

%calibrated data plotting

%figure
%plot(0:1/Fs:lenD/Fs-1/Fs,real(dataC),0:1/Fs:lenD/Fs-1/Fs,imag(dataC))
%title('Rx with offset')
%xlabel('time [s]')


N=wd*Fs/1000000 ; % conversion of window duration from miliseconds to samples

v=floor(lenD/N); % number of pulses recorded

y=zeros(N,v);

eRyy=0;
for i=0:v-1 %calculate autocorrelation matrix for v times
    
    y=dataC(i*N+1:i*N+N,1); % RX
    ryy= y*y';
    if(i>0)
        eRyy = (i*eRyy + ryy)/(i+1);
    end
end

 
%ERyy=sum(PA')/v; % Average the values 
Ryy= eRyy(:,1);
%ERyy = ERyy/ max(ERyy);
h = Ryy;



figure;
stem(0:1/Fs*1000000:N/Fs*1000000-1/Fs*1000000,abs(h)); % Plot the estimated impulse response

title('Estimated Impulse Response');
xlabel('time [us]')
grid;

%%%%%%%%%%%% GUARDAR RESPUESTA AL IMPULSO %%%%%%%%%%%%%%%%%

%guardarRespAlImpulso(abs(Ryx),savePath,Fs,wd,connectionType);


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%



%[maximum,pico]=max(abs(Ryx));
%pico=pico/Fs


%&&&&&&&&&&&&&&&&&&&&&&


%modu= conv(ruidoC,Ryx);

%lenM=size(modu(:,1));
%lenM=lenM(1,1);
%figure
%plot(0:1/Fs:lenM/Fs-1/Fs,real(modu),0:1/Fs:lenM/Fs-1/Fs,imag(modu))

File: modules/matrixDataProcess.py
#! /usr/bin/python
import sys
import numpy as np 
import matplotlib.pyplot as plt
from offcalc import offcalc
from SNRcalc import SNRcalc


#%%%%%%%%%5%%%%%%%%%%%%%
def matrixDataProcess(Fs , #Sample freq
	wd = 100,
	path = "/dev/shm/",
	offman = 0,
	plotMode=False):

	dataR = np.fromfile(path + "fdataR.dat",'float32')
	dataI = np.fromfile(path + "fdataI.dat",'float32')


	if len(dataR) != len(dataI):
		print '!'*40
		print "WARNING! SIZE ERROR OF RX SAMPLES"
		print "I:{}".format(len(dataR))
		print "Q:{}".format(len(dataI))
		print '!'*40
	
	dataC = dataR[0:int(2*Fs)] + 1j * dataI[0:int(2*Fs)]
	del dataR,dataI

	#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
	#Calibration processing

	calibrationOffset = 0.6 *Fs  #conversion from time to samples

	# signal calibration
	offset = offcalc(dataC[0:int(Fs*0.6)],Fs) + offman #received samples offset due to hardware & software lag [samples];
	snr = SNRcalc(dataC[0:int(Fs*0.8)],Fs,offset)

	print "SNR: {}".format(snr)

	dataC  = dataC [ offset : int(offset+Fs)  ]

	lenD=len(dataC)


	if 1 ==1 :
		N=int( wd*Fs/1000000) # conversion of window duration from miliseconds to samples

		v=int(np.floor(lenD/N)) # number of pulses recorded
		PA = np.zeros((N,v)) + 1j * np.zeros((N,v))
		

		for i in range(0,v):# Run cross correlation for v times
		    
			y=np.matrix((dataC[int(i*N)  : int(i*N+N)]))# RX
			y = y.transpose();

			ryy= y.dot(y.getH())# Cross correlation of the TX and RX conjugated data
			Ryy= ryy[:,1];

 			PA[:,[i]]=Ryy #Store the results in an array

		ERyy = PA.mean(axis=1) # Average the values 

		
		h= np.abs(ERyy)

	else:
		h = 0

	if plotMode :
		plt.stem(h)
		plt.show()
	else:
		return h


def main():
	Fs = 38e6 #Sample freq
	wd = 10; # window duration [us] for the correlation purpose
	path = "/dev/shm/"

	matrixDataProcess(Fs,wd,path,plotMode = True)


if __name__ == '__main__':
    main()
